File: frontend/src/lib/annotations.ts
/**
 * BODS statement annotations — "as filed" support.
 *
 * Phase 103 started emitting BODS `annotations` on statements: the Companies
 * House nature-of-control code behind a rendered `otherInfluenceOrControl`,
 * and a `commenting` note where a `birthDate` is imprecise because the
 * register published month and year only. Phase 107 carried them into the RDF
 * export. Until now nothing in the UI read them, so the register's own words
 * arrived in the browser and were discarded at the last step.
 *
 * The rule those annotations follow, from `bods/annotations.py`:
 * **the statement always carries the usable value; the annotation always
 * carries the register's words.** So "as filed" is a presentation swap, never
 * a different fact — the underlying statement is unchanged either way.
 *
 * Everything here is a pure function over the statement JSON so it can be
 * tested without a DOM (this codebase has no DOM test harness by choice).
 */

export interface BODSAnnotation {
  statementPointerTarget?: string;
  motivation?: string;
  description?: string;
  transformedContent?: string;
  creationDate?: string;
  createdBy?: { name?: string; uri?: string };
  url?: string;
}

/** RFC6901 pointer → path segments, unescaping `~1` → `/` and `~0` → `~`.
 *
 * Order matters: `~1` must be replaced before `~0`, or a literal `~1` encoded
 * as `~01` unescapes to `/` instead of `~1`. Same trap the Python side
 * documents.
 */
export function parsePointer(target: string): string[] {
  if (!target || target === "/") return [];
  return target
    .replace(/^\//, "")
    .split("/")
    .map((seg) => seg.replace(/~1/g, "/").replace(/~0/g, "~"));
}

function pathsEqual(a: string[], b: string[]): boolean {
  return a.length === b.length && a.every((seg, i) => seg === b[i]);
}

/** Every annotation on `stmt` whose pointer addresses exactly `path`.
 *
 * Exact match, not prefix: an annotation on `/recordDetails/interests/0/type`
 * describes that interest's type, and showing it against the whole interest —
 * or against a sibling interest — would attribute the register's words to
 * something it never said.
 */
export function annotationsAt(
  stmt: unknown,
  ...path: (string | number)[]
): BODSAnnotation[] {
  if (stmt == null || typeof stmt !== "object") return [];
  const raw = (stmt as Record<string, unknown>).annotations;
  if (!Array.isArray(raw)) return [];
  const wanted = path.map(String);
  return raw.filter((a): a is BODSAnnotation => {
    if (a == null || typeof a !== "object") return false;
    const target = (a as BODSAnnotation).statementPointerTarget;
    return typeof target === "string" && pathsEqual(parsePointer(target), wanted);
  });
}

/** The register's own words for a field, or "" if nothing was annotated.
 *
 * `description` is where the source's wording lives by construction — see the
 * one-rule note above. Multiple annotations on one field are joined rather
 * than silently dropping all but the first.
 */
export function asFiledText(annotations: BODSAnnotation[]): string {
  return annotations
    .map((a) => (a.description ?? "").trim())
    .filter(Boolean)
    .join(" · ");
}

/** How many fields across these statements carry an annotation.
 *
 * Drives whether the toggle renders at all: a control that changes nothing
 * is worse than no control, and most sources annotate nothing. Counts
 * distinct pointers per statement, so two annotations on one field count once.
 */
export function annotatedFieldCount(statements: unknown[]): number {
  let n = 0;
  for (const stmt of statements ?? []) {
    if (stmt == null || typeof stmt !== "object") continue;
    const raw = (stmt as Record<string, unknown>).annotations;
    if (!Array.isArray(raw)) continue;
    const seen = new Set<string>();
    for (const a of raw) {
      const target =
        a && typeof a === "object"
          ? (a as BODSAnnotation).statementPointerTarget
          : undefined;
      if (typeof target === "string" && target) seen.add(target);
    }
    n += seen.size;
  }
  return n;
}

/** Label for the toggle. States the destination, matching the graph's
 * "View as table" / "View as graph" convention rather than describing the
 * current state. */
export function asFiledToggleLabel(on: boolean, fieldCount: number): string {
  if (on) return "Show OpenCheck's reading";
  return fieldCount === 1
    ? "Show 1 field as filed"
    : `Show ${fieldCount} fields as filed`;
}

// ---------------------------------------------------------------------
// Shared toggle state
// ---------------------------------------------------------------------
//
// Deliberately module-scoped rather than per-card component state. A lookup
// renders many source cards; setting "as filed" on one and finding the next
// still in OpenCheck's vocabulary would read as a bug. One setting, applied
// wherever annotations exist, held for as long as the page lives.
//
// Not persisted to storage: the default should stay OpenCheck's reading,
// because that is the one that is always present and always machine-readable.

type Listener = () => void;

let asFiledOn = false;
const listeners = new Set<Listener>();

export function getAsFiled(): boolean {
  return asFiledOn;
}

export function setAsFiled(on: boolean): void {
  if (asFiledOn === on) return;
  asFiledOn = on;
  for (const l of listeners) l();
}

export function subscribeAsFiled(listener: Listener): () => void {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

/** Test-only: reset module state between cases. */
export function __resetAsFiled(): void {
  asFiledOn = false;
  listeners.clear();
}

File: backend/opencheck/bods/annotations.py
"""BODS annotations — recording what the register said, not just what we read.

OpenCheck transforms a lot on the way to BODS: register role labels become
``interestType`` codes, nature-of-control codes become English prose, dates
change format. The output carried none of that. A reader could see
``seniorManagingOfficial`` but not the Estonian label it came from, or a
``birthDate`` of ``1975-08`` without knowing whether the register published only
a month or OpenCheck truncated a full date.

BODS v0.4 has the right construct. Every statement may carry an ``annotations``
array; each entry pins an RFC6901 JSON Pointer at the fragment it describes:

    statementPointerTarget  required  e.g. "/recordDetails/interests/0/type"
    motivation              required  commenting | correcting | identifying
                                      | linking | transformation
    description             free text
    transformedContent      the representation after transformation
    creationDate            YYYY-MM-DD or ISO 8601
    createdBy               {name, uri}

One rule, applied everywhere
----------------------------
**The statement always carries the usable value; the annotation always carries
the register's words.**

``transformedContent`` is defined as the representation *after* transformation,
which read literally would put the original in the target field and the
normalised value in the annotation. That is unworkable for dates — BODS
constrains ``startDate`` and friends to ``YYYY-MM-DD``, so "01 November 2018"
cannot legally sit there — and inconsistent if applied only to free-string
fields. So the target field always holds the value a consumer should use, and
the annotation's ``description`` holds what the source actually said.
``transformedContent`` mirrors the emitted value, which keeps it truthful under
either reading. Worth raising upstream with Open Ownership; noted in the ticket.

What gets annotated
-------------------
Only transformations that are **lossy or non-obvious**. Annotating every field
would multiply bundle size for no gain — and the Render box is memory-bound
(see the Phase 88 OOM). An identity mapping is not worth a sentence; a
two-hop code translation that discards the register's own vocabulary is.
"""

from __future__ import annotations

from typing import Any, Iterable

#: BODS v0.4 annotation motivation codelist.
MOTIVATIONS: frozenset[str] = frozenset(
    {"commenting", "correcting", "identifying", "linking", "transformation"}
)

_PUBLISHER = {"name": "OpenCheck", "uri": "https://opencheck.world"}


def pointer(*parts: Any) -> str:
    """Build an RFC6901 JSON Pointer from path segments.

    Escaping matters: ``~`` becomes ``~0`` and ``/`` becomes ``~1``, or a
    field name containing either would silently address the wrong fragment.

    >>> pointer("recordDetails", "interests", 0, "type")
    '/recordDetails/interests/0/type'
    """
    escaped = [
        str(p).replace("~", "~0").replace("/", "~1") for p in parts
    ]
    return "/" + "/".join(escaped)


def transformation(
    target: str,
    description: str,
    *,
    transformed_content: str | None = None,
    creation_date: str | None = None,
) -> dict[str, Any]:
    """An annotation recording that OpenCheck transformed a source value.

    ``description`` should state what the source actually said, in the source's
    own words — that is the whole point of the record.
    """
    annotation: dict[str, Any] = {
        "statementPointerTarget": target,
        "motivation": "transformation",
        "description": description,
        "createdBy": dict(_PUBLISHER),
    }
    if transformed_content is not None:
        annotation["transformedContent"] = str(transformed_content)
    if creation_date:
        annotation["creationDate"] = creation_date
    return annotation


def commenting(
    target: str, description: str, *, creation_date: str | None = None
) -> dict[str, Any]:
    """An annotation that adds context without claiming a transformation.

    Used where the value is exactly what the source published but a consumer
    could reasonably misread it — an imprecise ``birthDate``, for instance,
    where BODS legitimately permits ``YYYY-MM`` and the reader cannot otherwise
    tell a privacy-limited register from a truncation on our side.
    """
    annotation: dict[str, Any] = {
        "statementPointerTarget": target,
        "motivation": "commenting",
        "description": description,
        "createdBy": dict(_PUBLISHER),
    }
    if creation_date:
        annotation["creationDate"] = creation_date
    return annotation


def annotate(
    statement: dict[str, Any], *annotations: dict[str, Any] | None
) -> dict[str, Any]:
    """Append annotations to *statement*, creating the array if needed.

    ``None`` entries are skipped so callers can pass a conditional result
    inline. Mutates and returns the statement.
    """
    for annotation in annotations:
        if annotation:
            statement.setdefault("annotations", []).append(annotation)
    return statement


def resolve_pointer(statement: dict[str, Any], target: str) -> Any:
    """Resolve an RFC6901 pointer against *statement*.

    Returns the addressed value, or raises KeyError/IndexError if the pointer
    does not resolve. Used by the tests to prove every annotation points at
    something that actually exists — pointers into arrays are fragile if
    statement construction order ever changes.
    """
    if target in ("", "/"):
        return statement
    node: Any = statement
    for raw in target.lstrip("/").split("/"):
        part = raw.replace("~1", "/").replace("~0", "~")
        if isinstance(node, list):
            node = node[int(part)]
        else:
            node = node[part]
    return node


# ----------------------------------------------------------------------
# Date precision
# ----------------------------------------------------------------------
# BODS is deliberately inconsistent here, and correctly so:
#
#   birthDate                       YYYY | YYYY-MM | YYYY-MM-DD all legal.
#                                   UK PSC publishes only month and year, on
#                                   purpose, for privacy. Rounding it would
#                                   fabricate precision AND defeat that.
#   foundingDate, dissolutionDate,  MUST be YYYY-MM-DD. Where the month or day
#   startDate, endDate,             is unknown the standard sanctions rounding
#   statementDate                   to the first of the month or year — but the
#                                   rounding is then invisible, so it should be
#                                   annotated.


def round_partial_date(value: str | None) -> tuple[str | None, str | None]:
    """Round a partial date to a full one, reporting what was rounded.

    Returns ``(iso_date, precision)`` where precision is ``"year"`` or
    ``"month"`` when rounding occurred, and ``None`` when the input was already
    a full date (or unusable). Follows the BODS rule: unknown day → first of
    month, unknown month → first of year.

    >>> round_partial_date("2022-03")
    ('2022-03-01', 'month')
    >>> round_partial_date("2022")
    ('2022-01-01', 'year')
    >>> round_partial_date("2022-03-15")
    ('2022-03-15', None)
    """
    if not value:
        return value, None
    text = str(value).strip()
    if len(text) == 4 and text.isdigit():
        return f"{text}-01-01", "year"
    if len(text) == 7 and text[4] == "-" and text[:4].isdigit() and text[5:].isdigit():
        return f"{text}-01", "month"
    return text, None


def date_rounding_annotation(
    target: str, original: str, precision: str, *, creation_date: str | None = None
) -> dict[str, Any]:
    """Record that a date was rounded because the source lacked precision.

    The BODS dates guidance asks publishers to communicate their rounding
    practice to data users. Prose in the docs satisfies that; this makes it
    machine-readable, so a consumer can tell a genuine 1 March from a rounded
    March without reading our documentation.
    """
    unit = "month and day" if precision == "year" else "day"
    return transformation(
        target,
        (
            f"Source supplied '{original}' — no {unit}. Rounded to the first "
            "day per the BODS dates guidance; the rounded portion is not a "
            "value the source stated."
        ),
        creation_date=creation_date,
    )


def validate_annotations(statement: dict[str, Any]) -> list[str]:
    """Return a list of problems with a statement's annotations, empty if fine.

    Checks the two things the schema cannot: that every ``motivation`` is in
    the codelist, and that every pointer actually resolves against this
    statement. A dangling pointer validates fine against the JSON schema and is
    useless to a consumer.
    """
    problems: list[str] = []
    for idx, annotation in enumerate(statement.get("annotations") or []):
        motivation = annotation.get("motivation")
        if motivation not in MOTIVATIONS:
            problems.append(
                f"annotations[{idx}]: motivation {motivation!r} not in the codelist"
            )
        target = annotation.get("statementPointerTarget")
        if not target:
            problems.append(f"annotations[{idx}]: missing statementPointerTarget")
            continue
        try:
            resolve_pointer(statement, target)
        except (KeyError, IndexError, ValueError):
            problems.append(
                f"annotations[{idx}]: pointer {target!r} does not resolve"
            )
    return problems


def validate_all(statements: Iterable[dict[str, Any]]) -> list[str]:
    """Run :func:`validate_annotations` across a bundle."""
    problems: list[str] = []
    for statement in statements:
        for problem in validate_annotations(statement):
            problems.append(f"{statement.get('statementId', '?')}: {problem}")
    return problems
